src/insexpand.c

```diff
@@ -3716,6 +3721,85 @@ get_next_default_completion(ins_compl_next_state_T *st, pos_T *start_pos)
 	if (ctrl_x_mode_line_or_eval() || (compl_cont_status & CONT_SOL))
 	    found_new_match = search_for_exact_line(st->ins_buf,
 			    st->cur_match_pos, compl_direction, compl_pattern);
+	else if (in_fuzzy)
+	{
+	    current_pos = *st->cur_match_pos;
+	    do {
+		if (st->ins_buf == curbuf)
+		     circly_end = *start_pos;
+		else
+		{
+		    circly_end.lnum = st->ins_buf->b_ml.ml_line_count;
+		    circly_end.col = 0;
+		    circly_end.coladd = 0;
+		}
+
+		// Check if looped around and back to start position
+		if (looped_around && EQUAL_POS(current_pos, circly_end))
+		    break;
+
+		// Get the current line buffer
+		if (current_pos.lnum >= 1
```

Comment:
Can `current_pos` be invalid here?
Wrap-around is _not_ controlled by `p_ws`.
`p_ic` is _not_ respected.
This code should be wrapped in a function (e.g. `search_for_fuzzy_match`) similar to `search_for_exact_line`, you can re-use most of it.